grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -0,0 +1,455 @@
+import org.apache.cassandra.stargate.exceptions.ReadTimeoutException;
+import org.apache.cassandra.stargate.exceptions.UnavailableException;
+import org.apache.cassandra.stargate.exceptions.WriteFailureException;
+import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
+
+/**
+ * @param <MessageT> the type of gRPC message being handled.
+ * @param <PreparedT> the persistence object resulting from the preparation of the query(ies).
+ */
+abstract class MessageHandler<MessageT extends GeneratedMessageV3, PreparedT> {
+
+  static final Metadata.Key<QueryOuterClass.Unavailable> UNAVAILABLE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.Unavailable.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteTimeout> WRITE_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadTimeout> READ_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadFailure> READ_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.FunctionFailure> FUNCTION_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.FunctionFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteFailure> WRITE_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.AlreadyExists> ALREADY_EXISTS_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.AlreadyExists.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.CasWriteUnknown> CAS_WRITE_UNKNOWN_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.CasWriteUnknown.getDefaultInstance());
+
+  protected static final ConsistencyLevel DEFAULT_TRACING_CONSISTENCY = ConsistencyLevel.ONE;
+
+  protected final MessageT message;
+  protected final Connection connection;
+  private final Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache;
+  protected final Persistence persistence;
+  private final StreamObserver<Response> responseObserver;
+
+  protected MessageHandler(
+      MessageT message,
+      Connection connection,
+      Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache,
+      Persistence persistence,
+      StreamObserver<Response> responseObserver) {
+    this.message = message;
+    this.connection = connection;
+    this.preparedCache = preparedCache;
+    this.persistence = persistence;
+    this.responseObserver = responseObserver;
+  }
+
+  void handle() {
+    CompletionStage<Result> resultFuture =
+        CompletableFuture.<Void>completedFuture(null)
+            .thenApply(this::invokeValidate)
+            .thenCompose(__ -> prepare(false))
+            .thenCompose(this::executePrepared);
+    resultFuture = handleUnprepared(resultFuture, this::reprepareAndRetry);
```

Comment:
`this::reprepareAndRetry` is always the second parameter. Maybe remove the second parameter?